cms/models/placeholdermodel.py

```diff
@@ -738,14 +738,51 @@ def _shift_plugin_positions(self, language, start, offset=None):
     def _recalculate_plugin_positions(self, language):
         from cms.models import CMSPlugin
         cursor = CMSPlugin._get_database_cursor('write')
-        sql = (
-            'UPDATE {0} '
-            'SET position = RowNbrs.RowNbr '
-            'FROM ('
-            'SELECT  ID, ROW_NUMBER() OVER (ORDER BY position) AS RowNbr '
-            'FROM {0} WHERE placeholder_id=%s AND language=%s '
-            ') RowNbrs '
-            'WHERE {0}.id=RowNbrs.id'
-        )
+
+        if connection.vendor == 'sqlite':
+            sql = (
+                'CREATE TEMPORARY TABLE temp AS '
+                'SELECT ID, ('
+                'SELECT COUNT(*)+1 FROM {0} t WHERE '
+                'placeholder_id={0}.placeholder_id AND language={0}.language '
+                'AND {0}.position > t.position'
+                ') AS new_position '
+                'FROM {0} WHERE placeholder_id=%s AND language=%s'
+            )
+            sql = sql.format(connection.ops.quote_name(CMSPlugin._meta.db_table))
+            cursor.execute(sql, [self.pk, language])
+
+            sql = (
+                'UPDATE {0} '
+                'SET position = (SELECT new_position FROM temp WHERE id={0}.id) '
+                'WHERE placeholder_id=%s AND language=%s'
+            )
+            sql = sql.format(connection.ops.quote_name(CMSPlugin._meta.db_table))
+            cursor.execute(sql, [self.pk, language])
+
+            sql = 'DROP TABLE temp'
+            sql = sql.format(connection.ops.quote_name(CMSPlugin._meta.db_table))
+            cursor.execute(sql)
+            return
+        elif connection.vendor == 'postgresql':
+            sql = (
+                'UPDATE {0} '
+                'SET position = RowNbrs.RowNbr '
+                'FROM ('
+                'SELECT  ID, ROW_NUMBER() OVER (ORDER BY position) AS RowNbr '
+                'FROM {0} WHERE placeholder_id=%s AND language=%s '
+                ') RowNbrs '
+                'WHERE {0}.id=RowNbrs.id'
+            )
+        else:
+            sql = (
```

Comment:
I assume we can't do this one also for sqlite?

Reply:
Unfortunately no, because sqlite runs `SELECT` inside the bracket every time after previous row was updated and recalculate the values, this results in bad position values. So we have to do this thing with `TEMPORARY TABLE`.